port `index_fill_` from TH to ATen.

nikitaved opened the pull request:
As per title. The port is based on TensorIterator.
Supports complex input.

Resolves https://github.com/pytorch/pytorch/issues/24714.
Resolves https://github.com/pytorch/pytorch/issues/24577.
Resolves https://github.com/pytorch/pytorch/issues/36328.
Possibly resolves https://github.com/pytorch/pytorch/issues/48230

nikitaved committed:
port index_fill from TH to ATen

nikitaved committed:
initial. has a bug regarding memory overlap

nikitaved committed:
resolve issue with memoverlap. The source is TensorIterator

nikitaved committed:
create empty stubs for kernels

nikitaved committed:
remove index_stride, not needed

nikitaved committed:
implement index_fill for CPU. TODO: index checks!

nikitaved committed:
push size into kernel

nikitaved commented:
As of now autograd tests will fail because of missing index checks. plus some input checks.

nikitaved committed:
push dim into kernels

nikitaved committed:
support for scalar self/index

nikitaved committed:
add comment about why memory overlap check is unset

nikitaved committed:
add CUDA support

nikitaved committed:
add check for index to be of type Long

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved committed:
minor

nikitaved committed:
minor

nikitaved committed:
kill TH code

nikitaved committed:
enable complex autograd tests

nikitaved committed:
modify map_arg to enable int64 Tensors

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2533,7 +2533,8 @@ def maybe_non_contig(tensor):
             if arg.dtype == torch.cfloat:
                 arg = arg.to(torch.cdouble)
-            if arg.is_complex() != dtype.is_complex:
+            # enable torch.int64 argument representing indices
+            if (arg.is_complex() != dtype.is_complex) and (arg.dtype != torch.int64):
```
@anjali411 , could you please tell whether it is fine to have this for now?

anjali411 replied:
Mmm I understand why you considered adding this, but I don't think we should add this. would you mind adding an `OpInfo` test for `torch.index_fill`?

nikitaved committed:
resolve merge conflict

nikitaved committed:
kill indexFill

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved commented:
Some benchmarks:
<details>
<summary>Test script</summary>

```python
from IPython import get_ipython
import torch

torch.manual_seed(13)
torch.set_num_threads(1)

ipython = get_ipython()

cpu = torch.device('cpu')
cuda = torch.device('cuda')

def run_test(ndims, size, index_len, device):
    print(f"ndims: {ndims}, tensor_size: {size}, index_len: {index_len}, device: {device}")
    x = torch.rand(*([size] * ndims), device=device)
    idx = (torch.rand(index_len, device=device) * size).long()
    for d in range(ndims):
        command = "x.index_fill(d, idx, -1)"
        if device == cuda:
            command = command + "; torch.cuda.synchronize()"
        ipython.magic(f"timeit {command}")
    print()

run_test(3, 1000, 10, cpu)
run_test(3, 1000, 100, cpu)
run_test(3, 1000, 1000, cpu)
run_test(2, 10000, 10000, cpu)

run_test(2, 10000, 10, cuda)
run_test(2, 10000, 100, cuda)
run_test(2, 10000, 1000, cuda)
run_test(2, 10000, 10000, cuda)


```

</details>

<details>
<summary>TH CUDA</summary>

```
ndims: 2, tensor_size: 10000, index_len: 10, device: cuda
2.76 ms ± 481 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
2.82 ms ± 578 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

ndims: 2, tensor_size: 10000, index_len: 100, device: cuda
2.77 ms ± 43.2 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
3.28 ms ± 276 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

ndims: 2, tensor_size: 10000, index_len: 1000, device: cuda
2.9 ms ± 180 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.58 ms ± 1.79 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

ndims: 2, tensor_size: 10000, index_len: 10000, device: cuda
4.39 ms ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
54.8 ms ± 37.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
</details>

<details>
<summary>Port CUDA</summary>

```
ndims: 2, tensor_size: 10000, index_len: 10, device: cuda                                                                                         
2.76 ms ± 249 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                            
2.82 ms ± 852 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                            
                                                                                                                                                  
ndims: 2, tensor_size: 10000, index_len: 100, device: cuda                                                                                        
2.77 ms ± 93.7 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                           
3.27 ms ± 216 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                            
                                                                                                                                                  
ndims: 2, tensor_size: 10000, index_len: 1000, device: cuda                                                                                       
2.89 ms ± 139 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                            
6.02 ms ± 2.44 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                           
                                                                                                                                                  
ndims: 2, tensor_size: 10000, index_len: 10000, device: cuda                                                                                      
4.05 ms ± 365 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)                                                                            
9.76 ms ± 3.59 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)        

```
</details>

<details>
<summary>TH CPU</summary>

```
ndims: 3, tensor_size: 1000, index_len: 10, device: cpu                                                                                           
1.08 s ± 2.81 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.08 s ± 2.71 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.16 s ± 1.34 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 3, tensor_size: 1000, index_len: 100, device: cpu
1.02 s ± 1.56 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.04 s ± 664 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.8 s ± 10.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 3, tensor_size: 1000, index_len: 1000, device: cpu
1.23 s ± 2.99 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.42 s ± 746 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
9.14 s ± 32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 2, tensor_size: 10000, index_len: 10000, device: cpu
139 ms ± 87.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
788 ms ± 321 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


```
</details>

<details>
<summary>Port CPU</summary>

```
1.11 s ± 19.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                               
1.16 s ± 30.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                               
1.22 s ± 14.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                               

ndims: 3, tensor_size: 1000, index_len: 100, device: cpu
1.2 s ± 17.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.27 s ± 32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.65 s ± 3.79 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 3, tensor_size: 1000, index_len: 1000, device: cpu
2.78 s ± 4.66 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
2.87 s ± 46.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
3.13 s ± 29 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 2, tensor_size: 10000, index_len: 10000, device: cpu
295 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
337 ms ± 2.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
</details>

When it comes to CUDA, the port appears to be superior all around. For CPU, however, it is not clear, especially with the test case `ndims: 3, tensor_size: 1000, index_len: 1000, device: cpu`.

ngimel commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -715,6 +716,77 @@ Tensor index_select_backward(const Tensor& grad, IntArrayRef self_sizes, int64_t
+    .add_input(index_restrided)
+    .build();
+
+  auto self_dim_size = self.size(dim);
```
unchecked access is slightly faster (self.sizes()[dim])

ngimel reviewed (commented):
This looks good! I don't see why cpu perf regresses - it looks like the traversal order is good, with localized writes. 
Please uncomment half and bfloat16 and complex tests in test_torch.py test_index_fill.

nikitaved committed:
uncomment blocked dtypes from test_torch.py for test_index_fill

nikitaved commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -715,6 +716,77 @@ Tensor index_select_backward(const Tensor& grad, IntArrayRef self_sizes, int64_t
+
+  // Prepare `index` for TensorIterator.
+  // It is restrided to be broadcastable over `self` in TensorIterator.
+  auto index_sizes = std::vector<int64_t>(self.dim(), 1);
```
No issues with having a hardcoded length?

nikitaved replied:
No issues with having a hardcoded dimensionality? We do not know `self.dim()` at compile time.

ngimel replied:
It grows to accomodate the required dimensionality (and then it's slow, but probably not slower than std::vector in the first place), but if you are within the specified max, should be faster, see e.g. #49497

nikitaved replied:
I see, but it does not have convenient Constructors as std::vector, however. I could set memory with some stl method I suppose. Shall I switch to SmallVector?

ngimel replied:
No, it's minor. If it doesn't work out of the box, then it's fine not using it.

nikitaved committed:
apply review remarks

nikitaved committed:
merge master

nikitaved commented:
I also do not know why regression happens, but comparing two test suites like 
```
ndims: 3, tensor_size: 1000, index_len: 100, device: cpu
1.03 s ± 757 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.04 s ± 539 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.81 s ± 3.41 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

ndims: 3, tensor_size: 1000, index_len: 1000, device: cpu
1.24 s ± 2.16 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.44 s ± 13.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
9.14 s ± 61.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
You could see that the running times do not change much even though the input size got increased by a factor of 10.
Also, this port does not cause issues like https://github.com/pytorch/pytorch/issues/48230, and it feels like the real issue is in the TH CPU implementation, but I do not know for sure...

ngimel commented on aten/src/ATen/native/cpu/IndexKernel.cpp:
```diff
@@ -131,6 +131,38 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
+        self_data[idx * self_dim_stride] = fill_val;
+
+        self_data_bytes += strides[0];
+        index_data_bytes += strides[1];
```
I got an idea that maybe if strides[1] is 0 (which is the case for indexing over non-last dimension) you should not be doing this, and instead get `idx` value outside of `elem` loop, and then just either move self_data_bytes forward or use self_data with increasing index. That might help compiler auto-vectorize this loop -currently it can't be done because `idx` is read every time, and compiler has no way of knowing that it will be the same.

ngimel replied:
Yep, taking `idx` out of the loop significantly improves perf for 1000,1000 case on my system, 3.7s->2.7s

ngimel replied:
Also valid index check for this case can be taken outside the loop, for additional perf gains.

nikitaved replied:
Done, indeed, now it is faster!

nikitaved committed:
take into account stride of the idx in the TensorIterator loop

nikitaved committed:
add OpInfo test for index_fill

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved committed:
revert test_autograd.py and gen_variable_type.py to the version from master

nikitaved committed:
restore complex checks in map_args

nikitaved committed:
minor typo fix

anjali411 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -470,6 +470,25 @@ def sample_inputs_index_select(op_info, device, dtype, requires_grad):
+    idx_nonctg.copy_(idx)
+    for d in range(t.dim()):
+        for tensor in [t, t01, t02, t12]:
+            samples.append(SampleInput((tensor, d, idx, -1)))
```
let's also add test for complex fill value

nikitaved replied:
Already added through OpInfo.

anjali411 replied:
wait I don't see an input with complex tensor and complex scalar value ...

anjali411 replied:
also let's update the doc to indicate `val` can be either complex or float https://pytorch.org/docs/stable/tensors.html#torch.Tensor.index_fill_

nikitaved replied:
Sorry, complex fill value, my bad.

nikitaved replied:
`jit_consistency` test fail when fill value is a complex number of the form `a+jb` with the following message:
```
                                            (*sample.input,) + sample.args,                                                                       
                                            sample.kwargs,                                                                                        
>                                           no_grad=(dtype not in dtypes_to_grad_check))

test/test_ops.py:293:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
torch/testing/_internal/common_jit.py:63: in check_against_reference
    outputs_test = self.runAndSaveRNG(func, nograd_inputs, kwargs)
torch/testing/_internal/common_jit.py:130: in runAndSaveRNG
    results = func(*inputs, **kwargs)
torch/testing/_internal/jit_metaprogramming_utils.py:298: in script_fn
    fn, tensors = gen_script_fn_and_args(method_name, func_type, *args, **kwargs)
torch/testing/_internal/jit_metaprogramming_utils.py:290: in gen_script_fn_and_args
    CU = torch.jit.CompilationUnit(script)
torch/jit/_script.py:1103: in __init__
    self.define(lang, _frames_up=_frames_up + 1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <torch.jit._script.CompilationUnit object at 0x7f2c1019f630>
lang = '\ndef the_method(i0, i1):\n    return torch.index_fill(i0, 0, i1, (-1+1j))\n'
rcb = <function createResolutionCallbackFromEnv.<locals>.<lambda> at 0x7f2be4c49400>, _frames_up = 1

    def define(self, lang, rcb=None, _frames_up=0):
        if not rcb:
            rcb = _jit_internal.createResolutionCallbackFromFrame(_frames_up + 1)
>       self._c.define(lang, rcb)
E       RuntimeError: expected ) but found 'ident' here:
E         File "<string>", line 3
E       
E       def the_method(i0, i1):
E           return torch.index_fill(i0, 0, i1, (-1+1j))
E                                                   ~ <--- HERE

torch/jit/_script.py:1108: RuntimeError
==========================================
```

anjali411 replied:
yeah that's expected. JIT doesn't know how to interpret complex literals right now. you can just skips this test in `OpInfo`, by adding:
```
skips=(
                 SkipInfo('TestCommon', 'test_variant_consistency_jit',
                          device_type='cuda', dtypes=[torch.cfloat, torch.cdouble]),
             )),
```

nikitaved replied:
I wrapped the complex scalar into a Tensor, there is an `index_fill` overload that accepts singleton Tensors.

anjali411 commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -170,6 +204,22 @@ static void index_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayR
+    iter.dtype(), "index_fill_cuda", [&] {
+    using dtype = OpaqueType<sizeof(scalar_t)>;
+    auto fill_val = source.to<scalar_t>();
+    auto fill_val_opaque = *reinterpret_cast<dtype*>(&fill_val);
```
why do we need this `reinterpret_cast` here?

nikitaved replied:
To `cast` to an Opaque type. Or could I use it directly?

anjali411 replied:
I don't think you need a `reinterpret_cast` here

nikitaved replied:
I do not think you can, not all dtypes could be converted to Opaque, the conversions are simply not defined. So, the only way is through casting the pointer, not the value.

anjali411 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -715,6 +716,75 @@ Tensor index_select_backward(const Tensor& grad, IntArrayRef self_sizes, int64_t
+      "Please clone() the tensor before performing this operation. "
+      "This also applies to advanced indexing e.g. tensor[mask] = scalar");
+  }
+
```
I think we should add a `TORCH_CHECK` to throw an error if we try to fill a complex value in a non-complex tensor or if we try to fill a floating point scalar in an int tensor. The latter would be a bc breaking change though:
```
>>> x = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.long)
>>> index = torch.tensor([0, 2])
>>> x.index_fill_(1, index, -1.4)
tensor([[-1,  2, -1],
        [-1,  5, -1],
        [-1,  8, -1]])
```
cc. @mruberry

anjali411 replied:
also 
```
>>> x.index_fill(1, index, -1.4)
tensor([[-1,  2, -1],
        [-1,  5, -1],
        [-1,  8, -1]])
```
I expected this to return a floating point tensor ...

nikitaved replied:
@anjali411 , you use the same `x` as from above? That one is of type `Long`. The scalar gets cast to an OpaqueType of lenght `sizeof(scalar_t)`, for `scalar_t == long` in this case.

ngimel replied:
`index_fill_` is an inplace operation, if called on an integral tensor it cannot return floating point tensor. Out-of-place variants, with rare exceptions, should have the same behavior as inplace. For the record, indexing assignment in numpy silently converts floating point to int, but errors out on complex to int or complex to float.

nikitaved replied:
Shall we adapt the numpy's behavior regarding complex fill value?

anjali411 replied:
yeah I think we should!

nikitaved replied:
How do I get the dtype from a Scalar then?

ngimel replied:
iirc there's `is_complex` function for Scalars

nikitaved replied:
Turns out there is `isComplex` method. So good the compiler was able to give the right suggestion from `is_complex` :)

ngimel commented on aten/src/ATen/native/cpu/IndexKernel.cpp:
```diff
@@ -131,6 +131,65 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
+        auto idx = *reinterpret_cast<int64_t*>(index_data_bytes);
+
+        if (idx < -self_dim_size || idx >= self_dim_size) {
+          TORCH_CHECK_INDEX(false,
```
nit: `TORCH_CHECK_INDEX(idx>=0 && idx < self_dim_size, "error msg")`. btw, I don't think we support negative indices here, as the access is just with `self_data[idx * self_dim_stride]`

nikitaved replied:
Good point!

nikitaved committed:
add support for neg idx

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved committed:
add tests with complex fill value

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved committed:
forbid casting complex Scalar to non-complex types

nikitaved committed:
Merge branch 'master' into nikved/index_fill_port

nikitaved commented:
@ngimel, @anjali411 , I hope I have addressed your comments. Please, let me know whether you are satisfied, or if there are any other issues that need resolution.

anjali411 reviewed (approved):
LGTM! thanks @nikitaved I'll wait for @ngimel to take a final look/

ngimel commented on aten/src/ATen/native/cpu/IndexKernel.cpp:
```diff
@@ -131,6 +131,70 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
+        if (idx < 0) {
+          idx += self_dim_size;
+        }
+        if (idx < -self_dim_size || idx >= self_dim_size) {
```
idx >=0 here and in all similar places, after you added self_dim_size to idx it has to be non-negative.

nikitaved replied:
Sorry, I will swap these two `if`s

ngimel commented on test/test_torch.py:
```diff
@@ -4304,13 +4304,13 @@ def test_func(slf, device, call_type):
 
     def test_index_fill(self, device):
         for dt in torch.testing.get_all_dtypes():
```
nit: instead of this loop, can you please use @dtypes decorator?

ngimel commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -715,6 +716,79 @@ Tensor index_select_backward(const Tensor& grad, IntArrayRef self_sizes, int64_t
+  }
+
+  if (!self.is_complex() && source.isComplex()) {
+    TORCH_CHECK(false, "Converting complex Scalar to non-complex type is forbidden");
```
Nit: can you insert `index_fill_` into error message text so that it's clear where it's thrown? The message is generic, and would be hard to debug if someone sees it.

nikitaved committed:
resolve conflicts with master

nikitaved committed:
address comments

nikitaved committed:
merge master

ngimel commented on aten/src/ATen/native/cpu/IndexKernel.cpp:
```diff
@@ -131,6 +131,70 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
+      if (idx < 0) {
+        idx += self_dim_size;
+      }
+      if (idx < -self_dim_size || idx >= self_dim_size) {
```
this one is still not swapped

nikitaved committed:
swap missed index sign check

anjali411 commented:
@nikitaved let me know once you've made the changes requested by @ngimel and I'll merge this.

ngimel commented:
It's ready!

nikitaved commented:
Hey, @anjali411! Yes, it is ready.

anjali411 commented:
@nikitaved could you please fix the lint errors?

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved committed:
fix flake8

nikitaved commented:
Hey, @anjali411 , sorry to bother you, is it going to get merged soon?

anjali411 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -716,6 +717,79 @@ Tensor index_select_backward(const Tensor& grad, IntArrayRef self_sizes, int64_t
+  }
+
+  // Handle the case when `self` is 0-dim
+  self = (0 == self.dim()) ? self.unsqueeze(-1) : self;
```
@nikitaved please update this to the inplace version `self.unsqueeze_(-1)`.

anjali411 commented:
@nikitaved sorry for keeping this PR on hold! just left a minor comment to fix an internal test failure and then it should be good to go

nikitaved committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nikved/index_fill_port

nikitaved committed:
use in-place unsqueeze for 0-dim tensors

nikitaved committed:
merge

nikitaved commented:
@anjali411 , sorry, there is something wrong with my history, I will try to fix that.

nikitaved committed:
Merge branch 'nikved/index_fill_port' of https://github.com/nikitaved/pytorch into nikved/index_fill_port
